python/paddle/fluid/tests/unittests/test_cumprod_op.py

```diff
@@ -0,0 +1,211 @@
+                        elem = dy[pos]
+                    else:
+                        elem = dy[pos] * y[index - inner_dim]
+                    if pos > index:
+                        for m in range(index + inner_dim, pos + inner_dim,
+                                       inner_dim):
+                            elem *= x[m]
+                    elif pos < index:
+                        elem = 0
+                    dx[index] += elem
+
+
+# test function.
+class TestCumprod(OpTest):
+    def init_params(self):
+        self.shape = (2, 3, 4, 5)
+        self.zero_nums = [0, 10, 20, 30, int(np.prod(self.shape))]
+
+    def init_dtype(self):
+        self.dtype = np.float64
+
+    def setUp(self):
+        paddle.enable_static()
+        self.init_params()
+        self.init_dtype()
+        self.op_type = "cumprod"
+        self.inputs = {'X': None}
+        self.outputs = {'Out': None}
+        self.attrs = {'dim': None}
+
+    def prepare_inputs_outputs_attrs(self, dim, zero_num):
+        self.x = np.random.random(self.shape).astype(self.dtype)
+        if zero_num > 0:
+            zero_num = min(zero_num, self.x.size)
+            shape = self.x.shape
+            self.x = self.x.flatten()
+            indices = random.sample(range(self.x.size), zero_num)
+            for i in indices:
+                self.x[i] = 0
+            self.x = np.reshape(self.x, self.shape)
+        self.out = np.cumprod(self.x, axis=dim)
+        self.inputs = {'X': OpTest.np_dtype_to_fluid_dtype(self.x)}
+        self.outputs = {'Out': self.out}
+        self.attrs = {'dim': dim}
+
+    def init_grad_input_output(self, dim):
+        reshape_x = self.x.reshape(self.x.size)
+        self.grad_out = np.ones(self.x.size, self.dtype)
+        self.grad_x = np.zeros(self.x.size, self.dtype)
+        out_data = self.out.reshape(self.x.size)
+        if self.dtype == np.complex128 or self.dtype == np.complex64:
+            reshape_x = np.conj(reshape_x)
+            out_data = np.conj(out_data)
+        cumprod_grad(reshape_x, out_data, self.grad_out, self.grad_x,
+                     self.shape, dim)
+        self.grad_x = self.grad_x.reshape(self.shape)
+        self.grad_out = self.grad_out.reshape(self.shape)
+
+    # test forward.
+    def test_check_output(self):
+        for dim in range(-len(self.shape), len(self.shape)):
+            for zero_num in self.zero_nums:
+                self.prepare_inputs_outputs_attrs(dim, zero_num)
+                self.check_output()
+
+    # test backward.
+    def test_check_grad(self):
+        for dim in range(-len(self.shape), len(self.shape)):
+            for zero_num in self.zero_nums:
+                self.prepare_inputs_outputs_attrs(dim, zero_num)
+                self.init_grad_input_output(dim)
+                self.check_grad(
+                    ['X'],
+                    'Out',
+                    user_defined_grads=[self.grad_x],
+                    user_defined_grad_outputs=[self.grad_out])
```

Comment:
I prefer to using `user_defined_grads` and `user_defined_grad_outputs` only in complex cases.